{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing loading and instantiation of workflows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------------------\n",
      "mixed_library_metrics.cwl STEPS\n",
      "-------------------------------------------------\n",
      "STEP Name: picard_collectmultiplemetrics\n",
      "-----------------------------------------------\n",
      "picard_collectmultiplemetrics INPUTS:\n",
      "INPUT Name: DB_SNP Source: WORKFLOW\n",
      "INPUT Name: INPUT Source: WORKFLOW\n",
      "INPUT Name: REFERENCE_SEQUENCE Source: WORKFLOW\n",
      "\n",
      "picard_collectmultiplemetrics OUTPUTS:\n",
      "OUTPUT Name: alignment_summary_metrics\n",
      "OUTPUT Name: bait_bias_detail_metrics\n",
      "OUTPUT Name: bait_bias_summary_metrics\n",
      "OUTPUT Name: base_distribution_by_cycle_metrics\n",
      "OUTPUT Name: gc_bias_detail_metrics\n",
      "OUTPUT Name: gc_bias_summary_metrics\n",
      "OUTPUT Name: insert_size_metrics\n",
      "OUTPUT Name: pre_adapter_detail_metrics\n",
      "OUTPUT Name: pre_adapter_summary_metrics\n",
      "OUTPUT Name: quality_by_cycle_metrics\n",
      "OUTPUT Name: quality_distribution_metrics\n",
      "OUTPUT Name: quality_yield_metrics\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: picard_collectmultiplemetrics_to_sqlite\n",
      "-----------------------------------------------\n",
      "picard_collectmultiplemetrics_to_sqlite INPUTS:\n",
      "INPUT Name: bam Source: WORKFLOW\n",
      "INPUT Name: fasta Source: WORKFLOW\n",
      "INPUT Name: input_state Source: WORKFLOW\n",
      "INPUT Name: uuid Source: WORKFLOW\n",
      "INPUT Name: vcf Source: WORKFLOW\n",
      "INPUT Name: alignment_summary_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: bait_bias_detail_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: bait_bias_summary_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: base_distribution_by_cycle_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: gc_bias_detail_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: gc_bias_summary_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: insert_size_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: pre_adapter_detail_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: pre_adapter_summary_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: quality_by_cycle_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: quality_distribution_metrics Source: picard_collectmultiplemetrics\n",
      "INPUT Name: quality_yield_metrics Source: picard_collectmultiplemetrics\n",
      "\n",
      "picard_collectmultiplemetrics_to_sqlite OUTPUTS:\n",
      "OUTPUT Name: log\n",
      "OUTPUT Name: sqlite\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: picard_collectoxogmetrics\n",
      "-----------------------------------------------\n",
      "picard_collectoxogmetrics INPUTS:\n",
      "INPUT Name: DB_SNP Source: WORKFLOW\n",
      "INPUT Name: INPUT Source: WORKFLOW\n",
      "INPUT Name: REFERENCE_SEQUENCE Source: WORKFLOW\n",
      "\n",
      "picard_collectoxogmetrics OUTPUTS:\n",
      "OUTPUT Name: OUTPUT\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: picard_collectoxogmetrics_to_sqlite\n",
      "-----------------------------------------------\n",
      "picard_collectoxogmetrics_to_sqlite INPUTS:\n",
      "INPUT Name: bam Source: WORKFLOW\n",
      "INPUT Name: fasta Source: WORKFLOW\n",
      "INPUT Name: input_state Source: WORKFLOW\n",
      "INPUT Name: metric_path Source: picard_collectoxogmetrics\n",
      "INPUT Name: uuid Source: WORKFLOW\n",
      "INPUT Name: vcf Source: WORKFLOW\n",
      "\n",
      "picard_collectoxogmetrics_to_sqlite OUTPUTS:\n",
      "OUTPUT Name: log\n",
      "OUTPUT Name: sqlite\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: picard_collectwgsmetrics\n",
      "-----------------------------------------------\n",
      "picard_collectwgsmetrics INPUTS:\n",
      "INPUT Name: INPUT Source: WORKFLOW\n",
      "INPUT Name: REFERENCE_SEQUENCE Source: WORKFLOW\n",
      "\n",
      "picard_collectwgsmetrics OUTPUTS:\n",
      "OUTPUT Name: OUTPUT\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: picard_collectwgsmetrics_to_sqlite\n",
      "-----------------------------------------------\n",
      "picard_collectwgsmetrics_to_sqlite INPUTS:\n",
      "INPUT Name: bam Source: WORKFLOW\n",
      "INPUT Name: fasta Source: WORKFLOW\n",
      "INPUT Name: input_state Source: WORKFLOW\n",
      "INPUT Name: metric_path Source: picard_collectwgsmetrics\n",
      "INPUT Name: uuid Source: WORKFLOW\n",
      "\n",
      "picard_collectwgsmetrics_to_sqlite OUTPUTS:\n",
      "OUTPUT Name: log\n",
      "OUTPUT Name: sqlite\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samtools_flagstat\n",
      "-----------------------------------------------\n",
      "samtools_flagstat INPUTS:\n",
      "INPUT Name: INPUT Source: WORKFLOW\n",
      "\n",
      "samtools_flagstat OUTPUTS:\n",
      "OUTPUT Name: OUTPUT\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samtools_flagstat_to_sqlite\n",
      "-----------------------------------------------\n",
      "samtools_flagstat_to_sqlite INPUTS:\n",
      "INPUT Name: bam Source: WORKFLOW\n",
      "INPUT Name: input_state Source: WORKFLOW\n",
      "INPUT Name: metric_path Source: samtools_flagstat\n",
      "INPUT Name: uuid Source: WORKFLOW\n",
      "\n",
      "samtools_flagstat_to_sqlite OUTPUTS:\n",
      "OUTPUT Name: sqlite\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samtools_idxstats\n",
      "-----------------------------------------------\n",
      "samtools_idxstats INPUTS:\n",
      "INPUT Name: INPUT Source: WORKFLOW\n",
      "\n",
      "samtools_idxstats OUTPUTS:\n",
      "OUTPUT Name: OUTPUT\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samtools_idxstats_to_sqlite\n",
      "-----------------------------------------------\n",
      "samtools_idxstats_to_sqlite INPUTS:\n",
      "INPUT Name: bam Source: WORKFLOW\n",
      "INPUT Name: input_state Source: WORKFLOW\n",
      "INPUT Name: metric_path Source: samtools_idxstats\n",
      "INPUT Name: uuid Source: WORKFLOW\n",
      "\n",
      "samtools_idxstats_to_sqlite OUTPUTS:\n",
      "OUTPUT Name: sqlite\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samtools_stats\n",
      "-----------------------------------------------\n",
      "samtools_stats INPUTS:\n",
      "INPUT Name: INPUT Source: WORKFLOW\n",
      "\n",
      "samtools_stats OUTPUTS:\n",
      "OUTPUT Name: OUTPUT\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samtools_stats_to_sqlite\n",
      "-----------------------------------------------\n",
      "samtools_stats_to_sqlite INPUTS:\n",
      "INPUT Name: bam Source: WORKFLOW\n",
      "INPUT Name: input_state Source: WORKFLOW\n",
      "INPUT Name: metric_path Source: samtools_stats\n",
      "INPUT Name: uuid Source: WORKFLOW\n",
      "\n",
      "samtools_stats_to_sqlite OUTPUTS:\n",
      "OUTPUT Name: sqlite\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: merge_sqlite\n",
      "-----------------------------------------------\n",
      "merge_sqlite INPUTS:\n",
      "INPUT Name: source_sqlite Type: list Source: WORKFLOW\n",
      "INPUT Name: uuid Source: WORKFLOW\n",
      "\n",
      "merge_sqlite OUTPUTS:\n",
      "OUTPUT Name: destination_sqlite\n",
      "OUTPUT Name: log\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "------------------------------------------------\n",
      "lobSTR.cwl STEPS\n",
      "-------------------------------------------------\n",
      "STEP Name: lobSTR\n",
      "-----------------------------------------------\n",
      "lobSTR INPUTS:\n",
      "INPUT Name: p1 Source: WORKFLOW\n",
      "INPUT Name: p2 Source: WORKFLOW\n",
      "INPUT Name: output_prefix Source: WORKFLOW\n",
      "INPUT Name: reference Source: WORKFLOW\n",
      "INPUT Name: rg-sample Source: WORKFLOW\n",
      "INPUT Name: rg-lib Source: WORKFLOW\n",
      "\n",
      "lobSTR OUTPUTS:\n",
      "OUTPUT Name: bam\n",
      "OUTPUT Name: bam_stats\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samsort\n",
      "-----------------------------------------------\n",
      "samsort INPUTS:\n",
      "INPUT Name: input Source: lobSTR\n",
      "INPUT Name: output_name Source: WORKFLOW\n",
      "\n",
      "samsort OUTPUTS:\n",
      "OUTPUT Name: output_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: samindex\n",
      "-----------------------------------------------\n",
      "samindex INPUTS:\n",
      "INPUT Name: input Source: samsort\n",
      "\n",
      "samindex OUTPUTS:\n",
      "OUTPUT Name: bam_with_bai\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: allelotype\n",
      "-----------------------------------------------\n",
      "allelotype INPUTS:\n",
      "INPUT Name: bam Source: samindex\n",
      "INPUT Name: reference Source: WORKFLOW\n",
      "INPUT Name: output_prefix Source: WORKFLOW\n",
      "INPUT Name: noise_model Source: WORKFLOW\n",
      "INPUT Name: strinfo Source: WORKFLOW\n",
      "\n",
      "allelotype OUTPUTS:\n",
      "OUTPUT Name: vcf\n",
      "OUTPUT Name: vcf_stats\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "lobSTR.cwl INPUTS:\n",
      "INPUT Name: p1 Type: File[]? Source: WORKFLOW\n",
      "INPUT Name: p2 Type: File[]? Source: WORKFLOW\n",
      "INPUT Name: output_prefix Type: string Source: WORKFLOW\n",
      "INPUT Name: reference Type: File Source: WORKFLOW\n",
      "INPUT Name: rg-sample Type: string Source: WORKFLOW\n",
      "INPUT Name: rg-lib Type: string Source: WORKFLOW\n",
      "INPUT Name: strinfo Type: File Source: WORKFLOW\n",
      "INPUT Name: noise_model Type: File Source: WORKFLOW\n",
      "\n",
      "v1.0/ATAC-seq_pipeline/03-map-se-blacklist-removal.cwl OUTPUTS:\n",
      "OUTPUT Name: output_pbc_files Type: File[]\n",
      "OUTPUT Name: output_read_count_mapped Type: File[]\n",
      "OUTPUT Name: output_data_sorted_dedup_bam_files Type: File[]\n",
      "OUTPUT Name: output_data_sorted_dups_marked_bam_files Type: File[]\n",
      "OUTPUT Name: output_picard_mark_duplicates_files Type: File[]\n",
      "OUTPUT Name: output_read_count_mapped_filtered Type: File[]\n",
      "OUTPUT Name: output_percentage_uniq_reads Type: File[]\n",
      "OUTPUT Name: output_bowtie_log Type: File[]\n",
      "OUTPUT Name: output_preseq_c_curve_files Type: File[]\n",
      "OUTPUT Name: output_percent_mitochondrial_reads Type: File[]\n",
      "\n",
      "------------------------------------------------\n",
      "v1.0/ATAC-seq_pipeline/03-map-se-blacklist-removal.cwl STEPS\n",
      "-------------------------------------------------\n",
      "STEP Name: extract_basename_1\n",
      "-----------------------------------------------\n",
      "extract_basename_1 INPUTS:\n",
      "INPUT Name: input_file Source: WORKFLOW\n",
      "\n",
      "extract_basename_1 OUTPUTS:\n",
      "OUTPUT Name: output_basename\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: extract_basename_2\n",
      "-----------------------------------------------\n",
      "extract_basename_2 INPUTS:\n",
      "INPUT Name: file_path Source: extract_basename_1\n",
      "\n",
      "extract_basename_2 OUTPUTS:\n",
      "OUTPUT Name: output_path\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: bowtie-se\n",
      "-----------------------------------------------\n",
      "bowtie-se INPUTS:\n",
      "INPUT Name: input_fastq_file Source: WORKFLOW\n",
      "INPUT Name: output_filename Source: extract_basename_2\n",
      "INPUT Name: v Source: WORKFLOW\n",
      "INPUT Name: X Source: WORKFLOW\n",
      "INPUT Name: genome_ref_first_index_file Source: WORKFLOW\n",
      "INPUT Name: nthreads Source: WORKFLOW\n",
      "\n",
      "bowtie-se OUTPUTS:\n",
      "OUTPUT Name: output_aligned_file\n",
      "OUTPUT Name: output_bowtie_log\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: sam2bam\n",
      "-----------------------------------------------\n",
      "sam2bam INPUTS:\n",
      "INPUT Name: nthreads Source: WORKFLOW\n",
      "INPUT Name: input_file Source: bowtie-se\n",
      "\n",
      "sam2bam OUTPUTS:\n",
      "OUTPUT Name: bam_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: sort_bams\n",
      "-----------------------------------------------\n",
      "sort_bams INPUTS:\n",
      "INPUT Name: nthreads Source: WORKFLOW\n",
      "INPUT Name: input_file Source: sam2bam\n",
      "\n",
      "sort_bams OUTPUTS:\n",
      "OUTPUT Name: sorted_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: index_bams\n",
      "-----------------------------------------------\n",
      "index_bams INPUTS:\n",
      "INPUT Name: input_file Source: sort_bams\n",
      "\n",
      "index_bams OUTPUTS:\n",
      "OUTPUT Name: indexed_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: filter-unmapped\n",
      "-----------------------------------------------\n",
      "filter-unmapped INPUTS:\n",
      "INPUT Name: output_filename Source: extract_basename_2\n",
      "INPUT Name: input_file Source: sort_bams\n",
      "\n",
      "filter-unmapped OUTPUTS:\n",
      "OUTPUT Name: filtered_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: filtered2sorted\n",
      "-----------------------------------------------\n",
      "filtered2sorted INPUTS:\n",
      "INPUT Name: nthreads Source: WORKFLOW\n",
      "INPUT Name: input_file Source: filter-unmapped\n",
      "\n",
      "filtered2sorted OUTPUTS:\n",
      "OUTPUT Name: sorted_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: index_filtered_bam\n",
      "-----------------------------------------------\n",
      "index_filtered_bam INPUTS:\n",
      "INPUT Name: input_file Source: filtered2sorted\n",
      "\n",
      "index_filtered_bam OUTPUTS:\n",
      "OUTPUT Name: indexed_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: preseq-c-curve\n",
      "-----------------------------------------------\n",
      "preseq-c-curve INPUTS:\n",
      "INPUT Name: input_sorted_file Source: filtered2sorted\n",
      "INPUT Name: output_file_basename Source: extract_basename_2\n",
      "\n",
      "preseq-c-curve OUTPUTS:\n",
      "OUTPUT Name: output_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: execute_pcr_bottleneck_coef\n",
      "-----------------------------------------------\n",
      "execute_pcr_bottleneck_coef INPUTS:\n",
      "INPUT Name: input_bam_files Source: filtered2sorted\n",
      "INPUT Name: genome_sizes Source: WORKFLOW\n",
      "INPUT Name: input_output_filenames Source: extract_basename_2\n",
      "\n",
      "execute_pcr_bottleneck_coef OUTPUTS:\n",
      "OUTPUT Name: pbc_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: remove_encode_blacklist\n",
      "-----------------------------------------------\n",
      "remove_encode_blacklist INPUTS:\n",
      "INPUT Name: a Source: filtered2sorted\n",
      "INPUT Name: b Source: WORKFLOW\n",
      "INPUT Name: output_basename_file Source: extract_basename_2\n",
      "INPUT Name: v Source: WORKFLOW\n",
      "\n",
      "remove_encode_blacklist OUTPUTS:\n",
      "OUTPUT Name: file_wo_blacklist_regions\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: masked_file_basename\n",
      "-----------------------------------------------\n",
      "masked_file_basename INPUTS:\n",
      "INPUT Name: input_file Source: remove_encode_blacklist\n",
      "\n",
      "masked_file_basename OUTPUTS:\n",
      "OUTPUT Name: output_basename\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: sort_masked_bams\n",
      "-----------------------------------------------\n",
      "sort_masked_bams INPUTS:\n",
      "INPUT Name: nthreads Source: WORKFLOW\n",
      "INPUT Name: input_file Source: remove_encode_blacklist\n",
      "\n",
      "sort_masked_bams OUTPUTS:\n",
      "OUTPUT Name: sorted_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: index_masked_bams\n",
      "-----------------------------------------------\n",
      "index_masked_bams INPUTS:\n",
      "INPUT Name: input_file Source: sort_masked_bams\n",
      "\n",
      "index_masked_bams OUTPUTS:\n",
      "OUTPUT Name: indexed_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: mark_duplicates\n",
      "-----------------------------------------------\n",
      "mark_duplicates INPUTS:\n",
      "INPUT Name: input_file Source: index_masked_bams\n",
      "INPUT Name: java_opts Source: WORKFLOW\n",
      "INPUT Name: picard_jar_path Source: WORKFLOW\n",
      "INPUT Name: output_filename Source: masked_file_basename\n",
      "INPUT Name: output_suffix Source: WORKFLOW\n",
      "\n",
      "mark_duplicates OUTPUTS:\n",
      "OUTPUT Name: output_metrics_file\n",
      "OUTPUT Name: output_dedup_bam_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: sort_dups_marked_bams\n",
      "-----------------------------------------------\n",
      "sort_dups_marked_bams INPUTS:\n",
      "INPUT Name: nthreads Source: WORKFLOW\n",
      "INPUT Name: input_file Source: mark_duplicates\n",
      "INPUT Name: suffix Source: WORKFLOW\n",
      "\n",
      "sort_dups_marked_bams OUTPUTS:\n",
      "OUTPUT Name: sorted_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: index_dups_marked_bams\n",
      "-----------------------------------------------\n",
      "index_dups_marked_bams INPUTS:\n",
      "INPUT Name: input_file Source: sort_dups_marked_bams\n",
      "\n",
      "index_dups_marked_bams OUTPUTS:\n",
      "OUTPUT Name: indexed_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: remove_duplicates\n",
      "-----------------------------------------------\n",
      "remove_duplicates INPUTS:\n",
      "INPUT Name: input_file Source: index_dups_marked_bams\n",
      "INPUT Name: F Source: WORKFLOW\n",
      "INPUT Name: suffix Source: WORKFLOW\n",
      "INPUT Name: b Source: WORKFLOW\n",
      "INPUT Name: outfile_name Source: WORKFLOW\n",
      "\n",
      "remove_duplicates OUTPUTS:\n",
      "OUTPUT Name: outfile\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: sort_dedup_bams\n",
      "-----------------------------------------------\n",
      "sort_dedup_bams INPUTS:\n",
      "INPUT Name: nthreads Source: WORKFLOW\n",
      "INPUT Name: input_file Source: remove_duplicates\n",
      "\n",
      "sort_dedup_bams OUTPUTS:\n",
      "OUTPUT Name: sorted_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: index_dedup_bams\n",
      "-----------------------------------------------\n",
      "index_dedup_bams INPUTS:\n",
      "INPUT Name: input_file Source: sort_dedup_bams\n",
      "\n",
      "index_dedup_bams OUTPUTS:\n",
      "OUTPUT Name: indexed_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: mapped_reads_count\n",
      "-----------------------------------------------\n",
      "mapped_reads_count INPUTS:\n",
      "INPUT Name: bowtie_log Source: bowtie-se\n",
      "\n",
      "mapped_reads_count OUTPUTS:\n",
      "OUTPUT Name: output\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: percent_uniq_reads\n",
      "-----------------------------------------------\n",
      "percent_uniq_reads INPUTS:\n",
      "INPUT Name: preseq_c_curve_outfile Source: preseq-c-curve\n",
      "\n",
      "percent_uniq_reads OUTPUTS:\n",
      "OUTPUT Name: output\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: mapped_filtered_reads_count\n",
      "-----------------------------------------------\n",
      "mapped_filtered_reads_count INPUTS:\n",
      "INPUT Name: output_suffix Source: WORKFLOW\n",
      "INPUT Name: input_bam_file Source: sort_dedup_bams\n",
      "\n",
      "mapped_filtered_reads_count OUTPUTS:\n",
      "OUTPUT Name: output_read_count\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: bam_idxstats\n",
      "-----------------------------------------------\n",
      "bam_idxstats INPUTS:\n",
      "INPUT Name: bam Source: index_bams\n",
      "\n",
      "bam_idxstats OUTPUTS:\n",
      "OUTPUT Name: idxstats_file\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: percent_mitochondrial_reads\n",
      "-----------------------------------------------\n",
      "percent_mitochondrial_reads INPUTS:\n",
      "INPUT Name: idxstats Source: bam_idxstats\n",
      "INPUT Name: chrom Source: WORKFLOW\n",
      "INPUT Name: output_filename Source: WORKFLOW\n",
      "\n",
      "percent_mitochondrial_reads OUTPUTS:\n",
      "OUTPUT Name: percent_map\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "varscan/varscan.cwl OUTPUTS:\n",
      "OUTPUT Name: snvs Type: File\n",
      "OUTPUT Name: indels Type: File\n",
      "OUTPUT Name: somatic_hc_snvs Type: File\n",
      "OUTPUT Name: somatic_snvs Type: File\n",
      "OUTPUT Name: germline_hc_snvs Type: File\n",
      "OUTPUT Name: germline_snvs Type: File\n",
      "OUTPUT Name: loh_hc_snvs Type: File\n",
      "OUTPUT Name: loh_snvs Type: File\n",
      "OUTPUT Name: somatic_hc_indels Type: File\n",
      "OUTPUT Name: somatic_indels Type: File\n",
      "OUTPUT Name: germline_hc_indels Type: File\n",
      "OUTPUT Name: germline_indels Type: File\n",
      "OUTPUT Name: loh_hc_indels Type: File\n",
      "OUTPUT Name: loh_indels Type: File\n",
      "\n",
      "workflows/simple-workflow.cwl OUTPUTS:\n",
      "OUTPUT Name: egg-milk-mixture Type: File\n",
      "OUTPUT Name: final-batter Type: File\n",
      "OUTPUT Name: cake Type: string\n",
      "\n",
      "------------------------------------------------\n",
      "rRNA_selection.cwl STEPS\n",
      "-------------------------------------------------\n",
      "STEP Name: hmmsearch\n",
      "-----------------------------------------------\n",
      "hmmsearch INPUTS:\n",
      "INPUT Name: hmm_profiles Source: WORKFLOW\n",
      "INPUT Name: sequence_query Source: WORKFLOW\n",
      "INPUT Name: e_threshold Source: WORKFLOW\n",
      "\n",
      "hmmsearch OUTPUTS:\n",
      "OUTPUT Name: per_domain_summary\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: extract_coord_lines\n",
      "-----------------------------------------------\n",
      "extract_coord_lines INPUTS:\n",
      "INPUT Name: summary Source: hmmsearch\n",
      "\n",
      "extract_coord_lines OUTPUTS:\n",
      "OUTPUT Name: coord_lines\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: extract_coordinates\n",
      "-----------------------------------------------\n",
      "extract_coordinates INPUTS:\n",
      "INPUT Name: coordinate_lines Source: extract_coord_lines\n",
      "\n",
      "extract_coordinates OUTPUTS:\n",
      "OUTPUT Name: formatted_names_and_coords\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: fetch_aligned_sequences\n",
      "-----------------------------------------------\n",
      "fetch_aligned_sequences INPUTS:\n",
      "INPUT Name: indexed_sequences Source: WORKFLOW\n",
      "INPUT Name: names Source: extract_coordinates\n",
      "INPUT Name: names_contain_subseq_coords Source: WORKFLOW\n",
      "\n",
      "fetch_aligned_sequences OUTPUTS:\n",
      "OUTPUT Name: sequences\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: reformat_matches\n",
      "-----------------------------------------------\n",
      "reformat_matches INPUTS:\n",
      "INPUT Name: sequences Source: fetch_aligned_sequences\n",
      "INPUT Name: minimum_length Source: WORKFLOW\n",
      "\n",
      "reformat_matches OUTPUTS:\n",
      "OUTPUT Name: reformatted_sequences\n",
      "\n",
      "\n",
      "-------------------------------------------------\n",
      "STEP Name: sort_matches\n",
      "-----------------------------------------------\n",
      "sort_matches INPUTS:\n",
      "INPUT Name: sequences Source: reformat_matches\n",
      "\n",
      "sort_matches OUTPUTS:\n",
      "OUTPUT Name: rev_sorted_sequences\n",
      "\n",
      "\n",
      "-------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "import cwlBrowser as c\n",
    "import workflow\n",
    "import requests\n",
    "import yaml as y \n",
    "from IPython.display import SVG, display\n",
    "\n",
    "\n",
    "#file = open(\"lobSTR.cwl\", 'rb')\n",
    "#lobSTR0 = y.safe_load(file)\n",
    "#print (lobSTR0[\"steps\"])\n",
    "lobSTR1 = c.load(\"lobSTR.cwl\")\n",
    "metrics = c.load(\"mixed_library_metrics.cwl\")\n",
    "another = c.load(\"https://github.com/alexbarrera/GGR-cwl/blob/master/v1.0/ATAC-seq_pipeline/03-map-se-blacklist-removal.cwl\", \n",
    "                         link=True)\n",
    "genomics = c.load(\"https://github.com/hamid58b/cancer-genomics-workflow/blob/master/varscan/varscan.cwl\",\n",
    "                         link=True)\n",
    "mine = c.load(\"https://github.com/seanleken/workflowtest/blob/master/workflows/simple-workflow.cwl\", link=True)\n",
    "rrna = c.load(\"rRNA_selection.cwl\")\n",
    "\n",
    "metrics.printSteps()\n",
    "lobSTR1.printSteps()\n",
    "lobSTR1.printInputs()\n",
    "another.printOutputs()\n",
    "another.printSteps()\n",
    "genomics.printOutputs()\n",
    "mine.printOutputs()\n",
    "rrna.printSteps()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing and loading ID type workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cwlBrowser as c\n",
    "link1 = \"https://github.com/NCI-GDC/gdc-dnaseq-cwl/blob/3fa737f9162bbfbdbc243a3b1ff7c8e110875eba/workflows/mirnaseq/mixed_library_metrics.cwl\"\n",
    "link2 = \"https://github.com/hamid58b/cancer-genomics-workflow/blob/master/varscan/varscan.cwl\"\n",
    "#print (link1)\n",
    "varcan = c.load(link2, link=True)\n",
    "metric = c.load(link1, link=True)\n",
    "#print (link2)\n",
    "#some = \"master/ddd.cwl\"\n",
    "#some.find(\"master\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing numeric comparison of workflows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------\n",
      "--------- inputs comparison------------\n",
      "-------------------------------------\n",
      "v1.0/ATAC-seq_pipeline/03-map-se-blacklist-removal.cwl| no. of inputs: 7\n",
      "varscan/varscan.cwl| no. of inputs: 4\n",
      "Difference: 3\n",
      "-------------------------------------\n",
      "\n",
      "\n",
      "-------------------------------------\n",
      "--------- outputs comparison------------\n",
      "-------------------------------------\n",
      "workflows/simple-workflow.cwl| no. of outputs: 3\n",
      "lobSTR.cwl| no. of outputs: 4\n",
      "Difference: 1\n",
      "-------------------------------------\n",
      "\n",
      "\n",
      "-------------------------------------\n",
      "--------- steps comparison------------\n",
      "-------------------------------------\n",
      "mixed_library_metrics.cwl| no. of steps: 13\n",
      "v1.0/ATAC-seq_pipeline/03-map-se-blacklist-removal.cwl| no. of steps: 26\n",
      "Difference: 13\n",
      "-------------------------------------\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import cwlBrowser as c\n",
    "\n",
    "lobSTR1 = c.load(\"lobSTR.cwl\")\n",
    "metrics = c.load(\"mixed_library_metrics.cwl\")\n",
    "another = c.loadWithLink(\"https://github.com/alexbarrera/GGR-cwl/blob/master/v1.0/ATAC-seq_pipeline/03-map-se-blacklist-removal.cwl\")\n",
    "genomics = c.loadWithLink(\"https://github.com/hamid58b/cancer-genomics-workflow/blob/master/varscan/varscan.cwl\")\n",
    "mine = c.loadWithLink(\"https://github.com/seanleken/workflowtest/blob/master/workflows/simple-workflow.cwl\")\n",
    "\n",
    "c.compareNoOfInputs(another, genomics)\n",
    "c.compareNoOfOutputs(mine, lobSTR1)\n",
    "c.compareNoOfSteps(metrics, another)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing display of workflow in graph form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg height=\"795pt\" viewBox=\"0.00 0.00 1591.00 818.00\" width=\"1547pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g class=\"graph\" id=\"graph0\" transform=\"scale(.9722 .9722) rotate(0) translate(4 814)\">\n",
       "<title>workflow</title>\n",
       "<g class=\"cluster\" id=\"clust1\">\n",
       "<title>cluster_inputs</title>\n",
       "<polygon fill=\"transparent\" points=\"391,-710 391,-765 1109,-765 1109,-710 391,-710\" stroke=\"#000000\" stroke-dasharray=\"5,2\"/>\n",
       "<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"433.5\" y=\"-753\">Workflow Inputs</text>\n",
       "</g>\n",
       "<g class=\"cluster\" id=\"clust2\">\n",
       "<title>cluster_outputs</title>\n",
       "<polygon fill=\"transparent\" points=\"8,-8 8,-63 1575,-63 1575,-8 8,-8\" stroke=\"#000000\" stroke-dasharray=\"5,2\"/>\n",
       "<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"54\" y=\"-15\">Workflow Outputs</text>\n",
       "</g>\n",
       "<!-- picard_jar_path -->\n",
       "<g class=\"node\" id=\"node1\">\n",
       "<title>picard_jar_path</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"399.5,-718.5 399.5,-737.5 482.5,-737.5 482.5,-718.5 399.5,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"441\" y=\"-725.5\">picard_jar_path</text>\n",
       "</g>\n",
       "<!-- mark_duplicates -->\n",
       "<g class=\"node\" id=\"node34\">\n",
       "<title>mark_duplicates</title>\n",
       "<polygon fill=\"#fafad2\" points=\"529,-305.5 529,-324.5 617,-324.5 617,-305.5 529,-305.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"573\" y=\"-312.5\">mark_duplicates</text>\n",
       "</g>\n",
       "<!-- picard_jar_path&#45;&gt;mark_duplicates -->\n",
       "<g class=\"edge\" id=\"edge26\">\n",
       "<title>picard_jar_path-&gt;mark_duplicates</title>\n",
       "<path d=\"M447.4148,-718.2956C453.5905,-707.9987 462,-691.0337 462,-675 462,-675 462,-675 462,-540 462,-472.5751 500.1655,-404.1883 524,-350 527.442,-342.1745 526.7645,-338.8485 533,-333 534.8885,-331.2287 536.9809,-329.6105 539.1898,-328.1358\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"540.4647,-330.2283 545.2487,-324.5612 537.9748,-326.008 540.4647,-330.2283\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"491.5\" y=\"-515.6\">picard_jar_path</text>\n",
       "</g>\n",
       "<!-- input_fastq_files -->\n",
       "<g class=\"node\" id=\"node2\">\n",
       "<title>input_fastq_files</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"913.5,-718.5 913.5,-737.5 1000.5,-737.5 1000.5,-718.5 913.5,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"957\" y=\"-725.5\">input_fastq_files</text>\n",
       "</g>\n",
       "<!-- extract_basename_1 -->\n",
       "<g class=\"node\" id=\"node26\">\n",
       "<title>extract_basename_1</title>\n",
       "<polygon fill=\"#fafad2\" points=\"1009.5,-790.5 1009.5,-809.5 1116.5,-809.5 1116.5,-790.5 1009.5,-790.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1063\" y=\"-797.5\">extract_basename_1</text>\n",
       "</g>\n",
       "<!-- input_fastq_files&#45;&gt;extract_basename_1 -->\n",
       "<g class=\"edge\" id=\"edge14\">\n",
       "<title>input_fastq_files-&gt;extract_basename_1</title>\n",
       "<path d=\"M968.5577,-737.5058C978.1116,-745.2166 992.1462,-756.2095 1005,-765 1016.1988,-772.6587 1029.0823,-780.5221 1039.858,-786.8506\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1038.6483,-788.9813 1045.9304,-790.3862 1041.1139,-784.7467 1038.6483,-788.9813\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1047\" y=\"-775.6\">input_file</text>\n",
       "</g>\n",
       "<!-- bowtie&#45;se -->\n",
       "<g class=\"node\" id=\"node35\">\n",
       "<title>bowtie-se</title>\n",
       "<polygon fill=\"#fafad2\" points=\"931.5,-665.5 931.5,-684.5 990.5,-684.5 990.5,-665.5 931.5,-665.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"961\" y=\"-672.5\">bowtie-se</text>\n",
       "</g>\n",
       "<!-- input_fastq_files&#45;&gt;bowtie&#45;se -->\n",
       "<g class=\"edge\" id=\"edge30\">\n",
       "<title>input_fastq_files-&gt;bowtie-se</title>\n",
       "<path d=\"M957.7337,-718.279C958.2919,-710.8827 959.0737,-700.5239 959.7324,-691.7951\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"962.1871,-691.824 960.271,-684.6594 957.301,-691.4551 962.1871,-691.824\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"986\" y=\"-695.6\">input_fastq_file</text>\n",
       "</g>\n",
       "<!-- nthreads -->\n",
       "<g class=\"node\" id=\"node3\">\n",
       "<title>nthreads</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"717,-718.5 717,-737.5 771,-737.5 771,-718.5 717,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"744\" y=\"-725.5\">nthreads</text>\n",
       "</g>\n",
       "<!-- filtered2sorted -->\n",
       "<g class=\"node\" id=\"node19\">\n",
       "<title>filtered2sorted</title>\n",
       "<polygon fill=\"#fafad2\" points=\"919.5,-485.5 919.5,-504.5 996.5,-504.5 996.5,-485.5 919.5,-485.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"958\" y=\"-492.5\">filtered2sorted</text>\n",
       "</g>\n",
       "<!-- nthreads&#45;&gt;filtered2sorted -->\n",
       "<g class=\"edge\" id=\"edge3\">\n",
       "<title>nthreads-&gt;filtered2sorted</title>\n",
       "<path d=\"M745.6342,-718.1045C750.3751,-689.9411 764.3448,-610.8664 773,-603 792.9086,-584.9058 810.0131,-608.977 833,-595 842.271,-589.3629 839.3807,-582.7249 847,-575 874.103,-547.5212 911.5162,-522.6969 935.2342,-508.2269\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"936.5178,-510.3139 941.2474,-504.6013 933.9876,-506.1176 936.5178,-510.3139\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"789\" y=\"-605.6\">nthreads</text>\n",
       "</g>\n",
       "<!-- sort_masked_bams -->\n",
       "<g class=\"node\" id=\"node21\">\n",
       "<title>sort_masked_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"546,-395.5 546,-414.5 648,-414.5 648,-395.5 546,-395.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"597\" y=\"-402.5\">sort_masked_bams</text>\n",
       "</g>\n",
       "<!-- nthreads&#45;&gt;sort_masked_bams -->\n",
       "<g class=\"edge\" id=\"edge6\">\n",
       "<title>nthreads-&gt;sort_masked_bams</title>\n",
       "<path d=\"M728.4967,-718.3364C723.6257,-715.4811 718.1684,-712.4636 713,-710 673.5089,-691.1764 621,-718.7479 621,-675 621,-675 621,-675 621,-517.5 621,-482.6916 609.8747,-443.1319 602.7734,-421.4263\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"605.0931,-420.6376 600.5413,-414.7823 600.4482,-422.1981 605.0931,-420.6376\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"637\" y=\"-560.6\">nthreads</text>\n",
       "</g>\n",
       "<!-- sort_dups_marked_bams -->\n",
       "<g class=\"node\" id=\"node29\">\n",
       "<title>sort_dups_marked_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"443,-260.5 443,-279.5 569,-279.5 569,-260.5 443,-260.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"506\" y=\"-267.5\">sort_dups_marked_bams</text>\n",
       "</g>\n",
       "<!-- nthreads&#45;&gt;sort_dups_marked_bams -->\n",
       "<g class=\"edge\" id=\"edge17\">\n",
       "<title>nthreads-&gt;sort_dups_marked_bams</title>\n",
       "<path d=\"M730.4087,-718.3871C725.1984,-715.1755 719.0485,-711.9266 713,-710 655.9872,-691.8401 583,-734.8352 583,-675 583,-675 583,-675 583,-585 583,-560.2983 579.954,-554.1998 575,-530 569.9897,-505.5251 525.149,-340.3187 510.5625,-286.7396\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"512.8417,-285.7846 508.6382,-279.6745 508.1139,-287.0723 512.8417,-285.7846\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"584\" y=\"-493.1\">nthreads</text>\n",
       "</g>\n",
       "<!-- sam2bam -->\n",
       "<g class=\"node\" id=\"node31\">\n",
       "<title>sam2bam</title>\n",
       "<polygon fill=\"#fafad2\" points=\"777.5,-620.5 777.5,-639.5 836.5,-639.5 836.5,-620.5 777.5,-620.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"807\" y=\"-627.5\">sam2bam</text>\n",
       "</g>\n",
       "<!-- nthreads&#45;&gt;sam2bam -->\n",
       "<g class=\"edge\" id=\"edge21\">\n",
       "<title>nthreads-&gt;sam2bam</title>\n",
       "<path d=\"M750.3214,-718.3848C753.532,-713.4879 757.487,-707.4353 761,-702 773.4289,-682.7704 787.5935,-660.5535 796.882,-645.943\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"799.0968,-647.0256 800.7827,-639.8034 794.9609,-644.3979 799.0968,-647.0256\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"800\" y=\"-673.1\">nthreads</text>\n",
       "</g>\n",
       "<!-- sort_dedup_bams -->\n",
       "<g class=\"node\" id=\"node33\">\n",
       "<title>sort_dedup_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"446,-125.5 446,-144.5 540,-144.5 540,-125.5 446,-125.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"493\" y=\"-132.5\">sort_dedup_bams</text>\n",
       "</g>\n",
       "<!-- nthreads&#45;&gt;sort_dedup_bams -->\n",
       "<g class=\"edge\" id=\"edge25\">\n",
       "<title>nthreads-&gt;sort_dedup_bams</title>\n",
       "<path d=\"M743.7054,-718.2327C743.4104,-707.5993 743,-690.093 743,-675 743,-675 743,-675 743,-585 743,-522.9033 779,-512.0967 779,-450 779,-450 779,-450 779,-225 779,-176.9499 630.256,-151.3967 547.3069,-140.8823\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"547.4761,-138.4346 540.2271,-140.0011 546.8708,-143.2971 547.4761,-138.4346\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"795\" y=\"-425.6\">nthreads</text>\n",
       "</g>\n",
       "<!-- nthreads&#45;&gt;bowtie&#45;se -->\n",
       "<g class=\"edge\" id=\"edge32\">\n",
       "<title>nthreads-&gt;bowtie-se</title>\n",
       "<path d=\"M759.523,-718.3776C764.3949,-715.524 769.8476,-712.497 775,-710 784.2466,-705.5188 808.0464,-695.5506 818,-693 853.4175,-683.9242 894.9909,-679.3448 924.1546,-677.0857\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"924.4894,-679.5177 931.2905,-676.56 924.1294,-674.631 924.4894,-679.5177\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"834\" y=\"-695.6\">nthreads</text>\n",
       "</g>\n",
       "<!-- sort_bams -->\n",
       "<g class=\"node\" id=\"node37\">\n",
       "<title>sort_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"847,-575.5 847,-594.5 909,-594.5 909,-575.5 847,-575.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"878\" y=\"-582.5\">sort_bams</text>\n",
       "</g>\n",
       "<!-- nthreads&#45;&gt;sort_bams -->\n",
       "<g class=\"edge\" id=\"edge35\">\n",
       "<title>nthreads-&gt;sort_bams</title>\n",
       "<path d=\"M756.7387,-718.2168C766.9645,-710.6672 781.8942,-700.3114 796,-693 804.8673,-688.4038 809.3557,-691.4279 817,-685 831.643,-672.6871 829.0757,-664.3564 839,-648 848.7736,-631.892 860.13,-613.5932 868.0586,-600.8835\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"870.2356,-602.0228 871.8658,-594.7878 866.0796,-599.4271 870.2356,-602.0228\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"855\" y=\"-650.6\">nthreads</text>\n",
       "</g>\n",
       "<!-- ENCODE_blacklist_bedfile -->\n",
       "<g class=\"node\" id=\"node4\">\n",
       "<title>ENCODE_blacklist_bedfile</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"581.5,-718.5 581.5,-737.5 712.5,-737.5 712.5,-718.5 581.5,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"647\" y=\"-725.5\">ENCODE_blacklist_bedfile</text>\n",
       "</g>\n",
       "<!-- remove_encode_blacklist -->\n",
       "<g class=\"node\" id=\"node39\">\n",
       "<title>remove_encode_blacklist</title>\n",
       "<polygon fill=\"#fafad2\" points=\"644.5,-440.5 644.5,-459.5 771.5,-459.5 771.5,-440.5 644.5,-440.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"708\" y=\"-447.5\">remove_encode_blacklist</text>\n",
       "</g>\n",
       "<!-- ENCODE_blacklist_bedfile&#45;&gt;remove_encode_blacklist -->\n",
       "<g class=\"edge\" id=\"edge40\">\n",
       "<title>ENCODE_blacklist_bedfile-&gt;remove_encode_blacklist</title>\n",
       "<path d=\"M604.1577,-718.4082C594.8551,-714.6084 585.7828,-709.3274 579,-702 570.498,-692.8154 571,-687.5157 571,-675 571,-675 571,-675 571,-517.5 571,-502.7243 567.821,-495.7102 578,-485 587.4258,-475.0822 615.4205,-466.9155 643.0281,-460.9725\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"643.6763,-463.3401 650.0271,-459.5098 642.6739,-458.5437 643.6763,-463.3401\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"573.5\" y=\"-583.1\">b</text>\n",
       "</g>\n",
       "<!-- picard_java_opts -->\n",
       "<g class=\"node\" id=\"node5\">\n",
       "<title>picard_java_opts</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"487,-718.5 487,-737.5 577,-737.5 577,-718.5 487,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"532\" y=\"-725.5\">picard_java_opts</text>\n",
       "</g>\n",
       "<!-- picard_java_opts&#45;&gt;mark_duplicates -->\n",
       "<g class=\"edge\" id=\"edge27\">\n",
       "<title>picard_java_opts-&gt;mark_duplicates</title>\n",
       "<path d=\"M530.8215,-718.2483C529.6416,-707.6282 528,-690.1332 528,-675 528,-675 528,-675 528,-360 528,-346.9774 537.5393,-336.4071 547.8338,-328.7916\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"549.507,-330.6166 553.9159,-324.653 546.7504,-326.5655 549.507,-330.6166\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"545.5\" y=\"-515.6\">java_opts</text>\n",
       "</g>\n",
       "<!-- genome_sizes_file -->\n",
       "<g class=\"node\" id=\"node6\">\n",
       "<title>genome_sizes_file</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"1005,-718.5 1005,-737.5 1101,-737.5 1101,-718.5 1005,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1053\" y=\"-725.5\">genome_sizes_file</text>\n",
       "</g>\n",
       "<!-- execute_pcr_bottleneck_coef -->\n",
       "<g class=\"node\" id=\"node24\">\n",
       "<title>execute_pcr_bottleneck_coef</title>\n",
       "<polygon fill=\"#f3cea1\" points=\"1090.5,-440.5 1090.5,-459.5 1233.5,-459.5 1233.5,-440.5 1090.5,-440.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1162\" y=\"-447.5\">execute_pcr_bottleneck_coef</text>\n",
       "</g>\n",
       "<!-- genome_sizes_file&#45;&gt;execute_pcr_bottleneck_coef -->\n",
       "<g class=\"edge\" id=\"edge12\">\n",
       "<title>genome_sizes_file-&gt;execute_pcr_bottleneck_coef</title>\n",
       "<path d=\"M1067.8246,-718.4811C1073.4373,-714.1266 1079.3474,-708.4857 1083,-702 1101.1424,-669.7856 1089.1175,-656.1217 1097,-620 1101.4131,-599.777 1102.2768,-594.5766 1109,-575 1122.7383,-534.9972 1143.2564,-489.6601 1154.3003,-466.12\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1156.6319,-466.9201 1157.4056,-459.5442 1152.2011,-464.8277 1156.6319,-466.9201\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1135\" y=\"-583.1\">genome_sizes</text>\n",
       "</g>\n",
       "<!-- genome_ref_first_index_file -->\n",
       "<g class=\"node\" id=\"node7\">\n",
       "<title>genome_ref_first_index_file</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"775,-718.5 775,-737.5 909,-737.5 909,-718.5 775,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"842\" y=\"-725.5\">genome_ref_first_index_file</text>\n",
       "</g>\n",
       "<!-- genome_ref_first_index_file&#45;&gt;bowtie&#45;se -->\n",
       "<g class=\"edge\" id=\"edge33\">\n",
       "<title>genome_ref_first_index_file-&gt;bowtie-se</title>\n",
       "<path d=\"M844.08,-718.4273C846.3485,-710.3655 850.8051,-699.1867 859,-693 869.1557,-685.333 899.2578,-680.6418 924.1475,-677.9795\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"924.6248,-680.3937 931.3417,-677.2494 924.1301,-675.5187 924.6248,-680.3937\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"908.5\" y=\"-695.6\">genome_ref_first_index_file</text>\n",
       "</g>\n",
       "<!-- output_bowtie_log -->\n",
       "<g class=\"node\" id=\"node8\">\n",
       "<title>output_bowtie_log</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"941.5,-35.5 941.5,-54.5 1036.5,-54.5 1036.5,-35.5 941.5,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"989\" y=\"-42.5\">output_bowtie_log</text>\n",
       "</g>\n",
       "<!-- output_read_count_mapped -->\n",
       "<g class=\"node\" id=\"node9\">\n",
       "<title>output_read_count_mapped</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"1338.5,-35.5 1338.5,-54.5 1475.5,-54.5 1475.5,-35.5 1338.5,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1407\" y=\"-42.5\">output_read_count_mapped</text>\n",
       "</g>\n",
       "<!-- output_preseq_c_curve_files -->\n",
       "<g class=\"node\" id=\"node10\">\n",
       "<title>output_preseq_c_curve_files</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"1194,-35.5 1194,-54.5 1334,-54.5 1334,-35.5 1194,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1264\" y=\"-42.5\">output_preseq_c_curve_files</text>\n",
       "</g>\n",
       "<!-- output_pbc_files -->\n",
       "<g class=\"node\" id=\"node11\">\n",
       "<title>output_pbc_files</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"1479.5,-35.5 1479.5,-54.5 1566.5,-54.5 1566.5,-35.5 1479.5,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1523\" y=\"-42.5\">output_pbc_files</text>\n",
       "</g>\n",
       "<!-- output_data_sorted_dedup_bam_files -->\n",
       "<g class=\"node\" id=\"node12\">\n",
       "<title>output_data_sorted_dedup_bam_files</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"404,-35.5 404,-54.5 582,-54.5 582,-35.5 404,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"493\" y=\"-42.5\">output_data_sorted_dedup_bam_files</text>\n",
       "</g>\n",
       "<!-- output_read_count_mapped_filtered -->\n",
       "<g class=\"node\" id=\"node13\">\n",
       "<title>output_read_count_mapped_filtered</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"230,-35.5 230,-54.5 400,-54.5 400,-35.5 230,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"315\" y=\"-42.5\">output_read_count_mapped_filtered</text>\n",
       "</g>\n",
       "<!-- output_percent_mitochondrial_reads -->\n",
       "<g class=\"node\" id=\"node14\">\n",
       "<title>output_percent_mitochondrial_reads</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"764.5,-35.5 764.5,-54.5 937.5,-54.5 937.5,-35.5 764.5,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"851\" y=\"-42.5\">output_percent_mitochondrial_reads</text>\n",
       "</g>\n",
       "<!-- output_percentage_uniq_reads -->\n",
       "<g class=\"node\" id=\"node15\">\n",
       "<title>output_percentage_uniq_reads</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"1040.5,-35.5 1040.5,-54.5 1189.5,-54.5 1189.5,-35.5 1040.5,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1115\" y=\"-42.5\">output_percentage_uniq_reads</text>\n",
       "</g>\n",
       "<!-- output_data_sorted_dups_marked_bam_files -->\n",
       "<g class=\"node\" id=\"node16\">\n",
       "<title>output_data_sorted_dups_marked_bam_files</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"16,-35.5 16,-54.5 226,-54.5 226,-35.5 16,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"121\" y=\"-42.5\">output_data_sorted_dups_marked_bam_files</text>\n",
       "</g>\n",
       "<!-- output_picard_mark_duplicates_files -->\n",
       "<g class=\"node\" id=\"node17\">\n",
       "<title>output_picard_mark_duplicates_files</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"586,-35.5 586,-54.5 760,-54.5 760,-35.5 586,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"673\" y=\"-42.5\">output_picard_mark_duplicates_files</text>\n",
       "</g>\n",
       "<!-- filter&#45;unmapped -->\n",
       "<g class=\"node\" id=\"node18\">\n",
       "<title>filter-unmapped</title>\n",
       "<polygon fill=\"#fafad2\" points=\"916.5,-530.5 916.5,-549.5 999.5,-549.5 999.5,-530.5 916.5,-530.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"958\" y=\"-537.5\">filter-unmapped</text>\n",
       "</g>\n",
       "<!-- filter&#45;unmapped&#45;&gt;filtered2sorted -->\n",
       "<g class=\"edge\" id=\"edge4\">\n",
       "<title>filter-unmapped-&gt;filtered2sorted</title>\n",
       "<path d=\"M958,-530.4567C958,-524.9678 958,-517.9184 958,-511.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"960.4501,-511.5296 958,-504.5296 955.5501,-511.5296 960.4501,-511.5296\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"974\" y=\"-515.6\">input_file</text>\n",
       "</g>\n",
       "<!-- filtered2sorted&#45;&gt;execute_pcr_bottleneck_coef -->\n",
       "<g class=\"edge\" id=\"edge11\">\n",
       "<title>filtered2sorted-&gt;execute_pcr_bottleneck_coef</title>\n",
       "<path d=\"M996.851,-486.4299C1029.4727,-479.234 1076.3019,-468.904 1111.7685,-461.0805\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1112.3948,-463.4513 1118.7027,-459.5509 1111.3393,-458.6663 1112.3948,-463.4513\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1106\" y=\"-470.6\">input_bam_files</text>\n",
       "</g>\n",
       "<!-- index_filtered_bam -->\n",
       "<g class=\"node\" id=\"node28\">\n",
       "<title>index_filtered_bam</title>\n",
       "<polygon fill=\"#fafad2\" points=\"964.5,-440.5 964.5,-459.5 1061.5,-459.5 1061.5,-440.5 964.5,-440.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1013\" y=\"-447.5\">index_filtered_bam</text>\n",
       "</g>\n",
       "<!-- filtered2sorted&#45;&gt;index_filtered_bam -->\n",
       "<g class=\"edge\" id=\"edge16\">\n",
       "<title>filtered2sorted-&gt;index_filtered_bam</title>\n",
       "<path d=\"M969.6641,-485.4567C977.2477,-479.2519 987.2686,-471.0529 995.7639,-464.1023\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"997.4865,-465.8585 1001.3527,-459.5296 994.3836,-462.0661 997.4865,-465.8585\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1006\" y=\"-470.6\">input_file</text>\n",
       "</g>\n",
       "<!-- preseq&#45;c&#45;curve -->\n",
       "<g class=\"node\" id=\"node32\">\n",
       "<title>preseq-c-curve</title>\n",
       "<polygon fill=\"#fafad2\" points=\"1040.5,-125.5 1040.5,-144.5 1121.5,-144.5 1121.5,-125.5 1040.5,-125.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1081\" y=\"-132.5\">preseq-c-curve</text>\n",
       "</g>\n",
       "<!-- filtered2sorted&#45;&gt;preseq&#45;c&#45;curve -->\n",
       "<g class=\"edge\" id=\"edge22\">\n",
       "<title>filtered2sorted-&gt;preseq-c-curve</title>\n",
       "<path d=\"M958.9489,-485.2193C960.0878,-474.187 962.1998,-455.7313 965,-440 980.9948,-350.141 983.1541,-325.1144 1023,-243 1039.181,-209.6541 1060.1882,-171.6968 1071.9669,-150.8341\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1074.1297,-151.9868 1075.4491,-144.6887 1069.8665,-149.5711 1074.1297,-151.9868\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1024.5\" y=\"-313.1\">input_sorted_file</text>\n",
       "</g>\n",
       "<!-- filtered2sorted&#45;&gt;remove_encode_blacklist -->\n",
       "<g class=\"edge\" id=\"edge38\">\n",
       "<title>filtered2sorted-&gt;remove_encode_blacklist</title>\n",
       "<path d=\"M919.4854,-488.0674C879.0989,-480.7978 815.0874,-469.2757 768.0572,-460.8103\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"768.2774,-458.3606 760.9541,-459.5317 767.4093,-463.1831 768.2774,-458.3606\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"857.5\" y=\"-470.6\">a</text>\n",
       "</g>\n",
       "<!-- index_dedup_bams -->\n",
       "<g class=\"node\" id=\"node20\">\n",
       "<title>index_dedup_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"443,-80.5 443,-99.5 543,-99.5 543,-80.5 443,-80.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"493\" y=\"-87.5\">index_dedup_bams</text>\n",
       "</g>\n",
       "<!-- index_dedup_bams&#45;&gt;output_data_sorted_dedup_bam_files -->\n",
       "<g class=\"edge\" id=\"edge49\">\n",
       "<title>index_dedup_bams-&gt;output_data_sorted_dedup_bam_files</title>\n",
       "<path d=\"M493,-80.4567C493,-74.9678 493,-67.9184 493,-61.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"495.4501,-61.5296 493,-54.5296 490.5501,-61.5296 495.4501,-61.5296\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- index_masked_bams -->\n",
       "<g class=\"node\" id=\"node23\">\n",
       "<title>index_masked_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"537,-350.5 537,-369.5 645,-369.5 645,-350.5 537,-350.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"591\" y=\"-357.5\">index_masked_bams</text>\n",
       "</g>\n",
       "<!-- sort_masked_bams&#45;&gt;index_masked_bams -->\n",
       "<g class=\"edge\" id=\"edge9\">\n",
       "<title>sort_masked_bams-&gt;index_masked_bams</title>\n",
       "<path d=\"M595.7276,-395.4567C594.9957,-389.9678 594.0558,-382.9184 593.2071,-376.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"595.6244,-376.1444 592.2706,-369.5296 590.7673,-376.792 595.6244,-376.1444\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"610\" y=\"-380.6\">input_file</text>\n",
       "</g>\n",
       "<!-- index_bams -->\n",
       "<g class=\"node\" id=\"node22\">\n",
       "<title>index_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"817,-170.5 817,-189.5 885,-189.5 885,-170.5 817,-170.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"851\" y=\"-177.5\">index_bams</text>\n",
       "</g>\n",
       "<!-- bam_idxstats -->\n",
       "<g class=\"node\" id=\"node27\">\n",
       "<title>bam_idxstats</title>\n",
       "<polygon fill=\"#fafad2\" points=\"814,-125.5 814,-144.5 888,-144.5 888,-125.5 814,-125.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"851\" y=\"-132.5\">bam_idxstats</text>\n",
       "</g>\n",
       "<!-- index_bams&#45;&gt;bam_idxstats -->\n",
       "<g class=\"edge\" id=\"edge15\">\n",
       "<title>index_bams-&gt;bam_idxstats</title>\n",
       "<path d=\"M851,-170.4567C851,-164.9678 851,-157.9184 851,-151.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"853.4501,-151.5296 851,-144.5296 848.5501,-151.5296 853.4501,-151.5296\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"859\" y=\"-155.6\">bam</text>\n",
       "</g>\n",
       "<!-- index_masked_bams&#45;&gt;mark_duplicates -->\n",
       "<g class=\"edge\" id=\"edge29\">\n",
       "<title>index_masked_bams-&gt;mark_duplicates</title>\n",
       "<path d=\"M587.1827,-350.4567C584.9394,-344.8485 582.0445,-337.6112 579.4556,-331.139\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"581.6864,-330.119 576.8118,-324.5296 577.1369,-331.9389 581.6864,-330.119\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"599\" y=\"-335.6\">input_file</text>\n",
       "</g>\n",
       "<!-- execute_pcr_bottleneck_coef&#45;&gt;output_pbc_files -->\n",
       "<g class=\"edge\" id=\"edge48\">\n",
       "<title>execute_pcr_bottleneck_coef-&gt;output_pbc_files</title>\n",
       "<path d=\"M1211.1199,-440.3922C1228.5837,-433.8793 1244,-423.0252 1244,-405 1244,-405 1244,-405 1244,-90 1244,-64.0482 1450.797,-69.189 1476,-63 1482.099,-61.5023 1488.4781,-59.4267 1494.4993,-57.2158\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1495.7053,-59.3774 1501.3651,-54.5847 1493.9519,-54.8019 1495.7053,-59.3774\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- extract_basename_2 -->\n",
       "<g class=\"node\" id=\"node25\">\n",
       "<title>extract_basename_2</title>\n",
       "<polygon fill=\"#fafad2\" points=\"1116.5,-718.5 1116.5,-737.5 1223.5,-737.5 1223.5,-718.5 1116.5,-718.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1170\" y=\"-725.5\">extract_basename_2</text>\n",
       "</g>\n",
       "<!-- extract_basename_2&#45;&gt;filter&#45;unmapped -->\n",
       "<g class=\"edge\" id=\"edge1\">\n",
       "<title>extract_basename_2-&gt;filter-unmapped</title>\n",
       "<path d=\"M1163.8939,-718.4805C1150.7967,-698.5512 1118.3133,-651.6658 1083,-620 1051.8763,-592.091 1009.9933,-567.4238 983.4912,-553.0998\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"984.4293,-550.8233 977.1007,-549.6853 982.1201,-555.1451 984.4293,-550.8233\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1131\" y=\"-628.1\">output_filename</text>\n",
       "</g>\n",
       "<!-- extract_basename_2&#45;&gt;execute_pcr_bottleneck_coef -->\n",
       "<g class=\"edge\" id=\"edge10\">\n",
       "<title>extract_basename_2-&gt;execute_pcr_bottleneck_coef</title>\n",
       "<path d=\"M1168.8215,-718.2483C1167.6416,-707.6282 1166,-690.1332 1166,-675 1166,-675 1166,-675 1166,-495 1166,-485.6322 1165.1293,-475.2103 1164.2004,-466.757\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1166.6129,-466.296 1163.3542,-459.634 1161.7472,-466.8741 1166.6129,-466.296\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1207.5\" y=\"-583.1\">input_output_filenames</text>\n",
       "</g>\n",
       "<!-- extract_basename_2&#45;&gt;preseq&#45;c&#45;curve -->\n",
       "<g class=\"edge\" id=\"edge23\">\n",
       "<title>extract_basename_2-&gt;preseq-c-curve</title>\n",
       "<path d=\"M1212.748,-718.4259C1234.3317,-710.8022 1256,-697.5365 1256,-675 1256,-675 1256,-675 1256,-405 1256,-374.9967 1255,-367.5033 1255,-337.5 1255,-337.5 1255,-337.5 1255,-225 1255,-167.9936 1180.4064,-147.0543 1128.8997,-139.3907\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1129.0676,-136.9404 1121.7953,-138.3951 1128.3875,-141.793 1129.0676,-136.9404\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1295.5\" y=\"-425.6\">output_file_basename</text>\n",
       "</g>\n",
       "<!-- extract_basename_2&#45;&gt;bowtie&#45;se -->\n",
       "<g class=\"edge\" id=\"edge31\">\n",
       "<title>extract_basename_2-&gt;bowtie-se</title>\n",
       "<path d=\"M1141.8809,-718.429C1131.9109,-715.3392 1120.548,-712.1583 1110,-710 1078.8893,-703.6343 1069.9271,-709.2048 1039,-702 1027.4988,-699.3207 1025.147,-696.899 1014,-693 1008.2642,-690.9937 1002.1683,-688.8936 996.2304,-686.8652\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"996.7796,-684.4642 989.3635,-684.5269 995.2001,-689.1026 996.7796,-684.4642\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1068\" y=\"-695.6\">output_filename</text>\n",
       "</g>\n",
       "<!-- extract_basename_2&#45;&gt;remove_encode_blacklist -->\n",
       "<g class=\"edge\" id=\"edge39\">\n",
       "<title>extract_basename_2-&gt;remove_encode_blacklist</title>\n",
       "<path d=\"M1143.1489,-718.4515C1132.9046,-715.2057 1121.0363,-711.9174 1110,-710 1043.3494,-698.4203 1022.9678,-720.8568 958,-702 950.1645,-699.7257 949.8409,-695.2558 942,-693 924.1473,-687.8639 789.1383,-697.0502 775,-685 740.4369,-655.5414 754,-630.4139 754,-585 754,-585 754,-585 754,-495 754,-481.7088 744.0656,-471.0526 733.4302,-463.4527\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"734.7007,-461.3558 727.5081,-459.5476 732.0033,-465.4465 734.7007,-461.3558\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"793.5\" y=\"-583.1\">output_basename_file</text>\n",
       "</g>\n",
       "<!-- extract_basename_1&#45;&gt;extract_basename_2 -->\n",
       "<g class=\"edge\" id=\"edge13\">\n",
       "<title>extract_basename_1-&gt;extract_basename_2</title>\n",
       "<path d=\"M1077.4305,-790.2898C1096.1213,-777.7128 1128.8005,-755.723 1149.9097,-741.5187\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1151.3583,-743.497 1155.7981,-737.5564 1148.6227,-739.4317 1151.3583,-743.497\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1115.5\" y=\"-775.6\">file_path</text>\n",
       "</g>\n",
       "<!-- percent_mitochondrial_reads -->\n",
       "<g class=\"node\" id=\"node40\">\n",
       "<title>percent_mitochondrial_reads</title>\n",
       "<polygon fill=\"#fafad2\" points=\"780.5,-80.5 780.5,-99.5 921.5,-99.5 921.5,-80.5 780.5,-80.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"851\" y=\"-87.5\">percent_mitochondrial_reads</text>\n",
       "</g>\n",
       "<!-- bam_idxstats&#45;&gt;percent_mitochondrial_reads -->\n",
       "<g class=\"edge\" id=\"edge41\">\n",
       "<title>bam_idxstats-&gt;percent_mitochondrial_reads</title>\n",
       "<path d=\"M851,-125.4567C851,-119.9678 851,-112.9184 851,-106.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"853.4501,-106.5296 851,-99.5296 848.5501,-106.5296 853.4501,-106.5296\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"864.5\" y=\"-110.6\">idxstats</text>\n",
       "</g>\n",
       "<!-- index_dups_marked_bams -->\n",
       "<g class=\"node\" id=\"node38\">\n",
       "<title>index_dups_marked_bams</title>\n",
       "<polygon fill=\"#fafad2\" points=\"427,-215.5 427,-234.5 559,-234.5 559,-215.5 427,-215.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"493\" y=\"-222.5\">index_dups_marked_bams</text>\n",
       "</g>\n",
       "<!-- sort_dups_marked_bams&#45;&gt;index_dups_marked_bams -->\n",
       "<g class=\"edge\" id=\"edge37\">\n",
       "<title>sort_dups_marked_bams-&gt;index_dups_marked_bams</title>\n",
       "<path d=\"M503.243,-260.4567C501.6574,-254.9678 499.6209,-247.9184 497.782,-241.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"500.0496,-240.5746 495.753,-234.5296 495.3421,-241.9346 500.0496,-240.5746\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"516\" y=\"-245.6\">input_file</text>\n",
       "</g>\n",
       "<!-- masked_file_basename -->\n",
       "<g class=\"node\" id=\"node30\">\n",
       "<title>masked_file_basename</title>\n",
       "<polygon fill=\"#fafad2\" points=\"652.5,-395.5 652.5,-414.5 771.5,-414.5 771.5,-395.5 652.5,-395.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"712\" y=\"-402.5\">masked_file_basename</text>\n",
       "</g>\n",
       "<!-- masked_file_basename&#45;&gt;mark_duplicates -->\n",
       "<g class=\"edge\" id=\"edge28\">\n",
       "<title>masked_file_basename-&gt;mark_duplicates</title>\n",
       "<path d=\"M701.9348,-395.367C689.5804,-383.8143 667.7134,-364.2296 647,-350 634.0483,-341.1025 630.1558,-339.821 616,-333 612.0767,-331.1095 607.9028,-329.2185 603.7569,-327.4107\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"604.7047,-325.1514 597.3053,-324.6503 602.7772,-329.6564 604.7047,-325.1514\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"701\" y=\"-358.1\">output_filename</text>\n",
       "</g>\n",
       "<!-- sam2bam&#45;&gt;sort_bams -->\n",
       "<g class=\"edge\" id=\"edge36\">\n",
       "<title>sam2bam-&gt;sort_bams</title>\n",
       "<path d=\"M805.9492,-620.3944C805.8666,-614.6677 806.7987,-607.6341 811,-603 819.4618,-593.6664 826.5544,-597.2672 839.7306,-595.4187\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"840.3512,-597.7926 846.7202,-593.9929 839.3718,-592.9915 840.3512,-597.7926\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"827\" y=\"-605.6\">input_file</text>\n",
       "</g>\n",
       "<!-- preseq&#45;c&#45;curve&#45;&gt;output_preseq_c_curve_files -->\n",
       "<g class=\"edge\" id=\"edge47\">\n",
       "<title>preseq-c-curve-&gt;output_preseq_c_curve_files</title>\n",
       "<path d=\"M1102.8545,-125.4533C1106.5922,-123.0976 1110.1745,-120.2869 1113,-117 1124.4616,-103.6671 1113.8934,-91.7198 1127,-80 1128.7712,-78.4162 1178.5328,-65.9408 1217.5753,-56.3306\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1218.5139,-58.6229 1224.7267,-54.5728 1217.3442,-53.8645 1218.5139,-58.6229\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- percent_uniq_reads -->\n",
       "<g class=\"node\" id=\"node42\">\n",
       "<title>percent_uniq_reads</title>\n",
       "<polygon fill=\"#fafad2\" points=\"1008.5,-80.5 1008.5,-99.5 1109.5,-99.5 1109.5,-80.5 1008.5,-80.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1059\" y=\"-87.5\">percent_uniq_reads</text>\n",
       "</g>\n",
       "<!-- preseq&#45;c&#45;curve&#45;&gt;percent_uniq_reads -->\n",
       "<g class=\"edge\" id=\"edge43\">\n",
       "<title>preseq-c-curve-&gt;percent_uniq_reads</title>\n",
       "<path d=\"M1047.0081,-125.3239C1040.4757,-122.7811 1035.0278,-119.9261 1033,-117 1029.9246,-112.5623 1031.5737,-108.165 1035.2079,-104.2481\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1036.9412,-105.9895 1040.6667,-99.5767 1033.7553,-102.2665 1036.9412,-105.9895\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1074.5\" y=\"-110.6\">preseq_c_curve_outfile</text>\n",
       "</g>\n",
       "<!-- sort_dedup_bams&#45;&gt;index_dedup_bams -->\n",
       "<g class=\"edge\" id=\"edge5\">\n",
       "<title>sort_dedup_bams-&gt;index_dedup_bams</title>\n",
       "<path d=\"M493,-125.4567C493,-119.9678 493,-112.9184 493,-106.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"495.4501,-106.5296 493,-99.5296 490.5501,-106.5296 495.4501,-106.5296\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"509\" y=\"-110.6\">input_file</text>\n",
       "</g>\n",
       "<!-- mapped_filtered_reads_count -->\n",
       "<g class=\"node\" id=\"node43\">\n",
       "<title>mapped_filtered_reads_count</title>\n",
       "<polygon fill=\"#fafad2\" points=\"269.5,-80.5 269.5,-99.5 412.5,-99.5 412.5,-80.5 269.5,-80.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"341\" y=\"-87.5\">mapped_filtered_reads_count</text>\n",
       "</g>\n",
       "<!-- sort_dedup_bams&#45;&gt;mapped_filtered_reads_count -->\n",
       "<g class=\"edge\" id=\"edge44\">\n",
       "<title>sort_dedup_bams-&gt;mapped_filtered_reads_count</title>\n",
       "<path d=\"M460.7647,-125.4567C437.1865,-118.4763 405.0833,-108.972 380.0493,-101.5606\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"380.5964,-99.1676 373.1888,-99.5296 379.2053,-103.866 380.5964,-99.1676\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"456\" y=\"-110.6\">input_bam_file</text>\n",
       "</g>\n",
       "<!-- mark_duplicates&#45;&gt;output_picard_mark_duplicates_files -->\n",
       "<g class=\"edge\" id=\"edge54\">\n",
       "<title>mark_duplicates-&gt;output_picard_mark_duplicates_files</title>\n",
       "<path d=\"M617.2485,-313.5733C674.7817,-310.4894 768,-300.7484 768,-270 768,-270 768,-270 768,-90 768,-77.3517 740.1482,-65.2075 714.6058,-56.7296\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"715.2492,-54.363 707.8351,-54.547 713.7457,-59.0266 715.2492,-54.363\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- mark_duplicates&#45;&gt;sort_dups_marked_bams -->\n",
       "<g class=\"edge\" id=\"edge18\">\n",
       "<title>mark_duplicates-&gt;sort_dups_marked_bams</title>\n",
       "<path d=\"M558.791,-305.4567C549.2863,-299.0729 536.6387,-290.5783 526.1059,-283.504\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"527.3655,-281.3987 520.1885,-279.5296 524.6335,-285.4664 527.3655,-281.3987\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"561\" y=\"-290.6\">input_file</text>\n",
       "</g>\n",
       "<!-- bowtie&#45;se&#45;&gt;output_bowtie_log -->\n",
       "<g class=\"edge\" id=\"edge45\">\n",
       "<title>bowtie-se-&gt;output_bowtie_log</title>\n",
       "<path d=\"M990.5215,-668.6356C1022.1928,-660.934 1068,-646.8776 1068,-630 1068,-630 1068,-630 1068,-472.5 1068,-397.7167 1069.4085,-377.7289 1052,-305 1035.6022,-236.4935 994,-227.9416 994,-157.5 994,-157.5 994,-157.5 994,-90 994,-80.6115 992.9116,-70.1883 991.7505,-61.7394\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"994.1452,-61.1854 990.6928,-54.6216 989.2984,-61.9057 994.1452,-61.1854\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- bowtie&#45;se&#45;&gt;sam2bam -->\n",
       "<g class=\"edge\" id=\"edge20\">\n",
       "<title>bowtie-se-&gt;sam2bam</title>\n",
       "<path d=\"M931.3871,-665.8188C914.0995,-660.5083 891.8495,-653.7583 872,-648 862.8593,-645.3483 852.9704,-642.5659 843.7299,-640.0031\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"844.2129,-637.5948 836.8132,-638.0919 842.9078,-642.3178 844.2129,-637.5948\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"917\" y=\"-650.6\">input_file</text>\n",
       "</g>\n",
       "<!-- mapped_reads_count -->\n",
       "<g class=\"node\" id=\"node41\">\n",
       "<title>mapped_reads_count</title>\n",
       "<polygon fill=\"#fafad2\" points=\"1127,-80.5 1127,-99.5 1237,-99.5 1237,-80.5 1127,-80.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"1182\" y=\"-87.5\">mapped_reads_count</text>\n",
       "</g>\n",
       "<!-- bowtie&#45;se&#45;&gt;mapped_reads_count -->\n",
       "<g class=\"edge\" id=\"edge42\">\n",
       "<title>bowtie-se-&gt;mapped_reads_count</title>\n",
       "<path d=\"M990.7134,-673.7762C1019.4752,-671.9752 1060.3833,-667.5846 1072,-657 1081.1634,-648.6507 1079,-642.3967 1079,-630 1079,-630 1079,-630 1079,-450 1079,-312.1657 1149.6547,-155.7004 1173.9272,-106.0156\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1176.13,-107.088 1177.0321,-99.7266 1171.7363,-104.9187 1176.13,-107.088\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"1103.5\" y=\"-380.6\">bowtie_log</text>\n",
       "</g>\n",
       "<!-- remove_duplicates -->\n",
       "<g class=\"node\" id=\"node36\">\n",
       "<title>remove_duplicates</title>\n",
       "<polygon fill=\"#fafad2\" points=\"444,-170.5 444,-189.5 542,-189.5 542,-170.5 444,-170.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"493\" y=\"-177.5\">remove_duplicates</text>\n",
       "</g>\n",
       "<!-- remove_duplicates&#45;&gt;sort_dedup_bams -->\n",
       "<g class=\"edge\" id=\"edge24\">\n",
       "<title>remove_duplicates-&gt;sort_dedup_bams</title>\n",
       "<path d=\"M493,-170.4567C493,-164.9678 493,-157.9184 493,-151.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"495.4501,-151.5296 493,-144.5296 490.5501,-151.5296 495.4501,-151.5296\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"509\" y=\"-155.6\">input_file</text>\n",
       "</g>\n",
       "<!-- sort_bams&#45;&gt;filter&#45;unmapped -->\n",
       "<g class=\"edge\" id=\"edge2\">\n",
       "<title>sort_bams-&gt;filter-unmapped</title>\n",
       "<path d=\"M894.9659,-575.4567C906.527,-568.9536 921.9823,-560.26 934.6955,-553.1088\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"936.1586,-555.0968 941.0585,-549.5296 933.7563,-550.8261 936.1586,-555.0968\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"941\" y=\"-560.6\">input_file</text>\n",
       "</g>\n",
       "<!-- sort_bams&#45;&gt;index_bams -->\n",
       "<g class=\"edge\" id=\"edge8\">\n",
       "<title>sort_bams-&gt;index_bams</title>\n",
       "<path d=\"M867.9462,-575.2728C860.2697,-566.7669 851,-553.6663 851,-540 851,-540 851,-540 851,-225 851,-215.6689 851,-205.2495 851,-196.7884\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"853.4501,-196.6559 851,-189.656 848.5501,-196.656 853.4501,-196.6559\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"867\" y=\"-380.6\">input_file</text>\n",
       "</g>\n",
       "<!-- index_dups_marked_bams&#45;&gt;output_data_sorted_dups_marked_bam_files -->\n",
       "<g class=\"edge\" id=\"edge53\">\n",
       "<title>index_dups_marked_bams-&gt;output_data_sorted_dups_marked_bam_files</title>\n",
       "<path d=\"M426.7924,-222.0301C334.6683,-216.9954 179,-204.8163 179,-180 179,-180 179,-180 179,-90 179,-75.5427 167.6116,-65.0621 154.9276,-57.8726\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"155.9821,-55.6595 148.6411,-54.6046 153.7219,-60.0072 155.9821,-55.6595\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- index_dups_marked_bams&#45;&gt;remove_duplicates -->\n",
       "<g class=\"edge\" id=\"edge34\">\n",
       "<title>index_dups_marked_bams-&gt;remove_duplicates</title>\n",
       "<path d=\"M493,-215.4567C493,-209.9678 493,-202.9184 493,-196.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"495.4501,-196.5296 493,-189.5296 490.5501,-196.5296 495.4501,-196.5296\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"509\" y=\"-200.6\">input_file</text>\n",
       "</g>\n",
       "<!-- remove_encode_blacklist&#45;&gt;sort_masked_bams -->\n",
       "<g class=\"edge\" id=\"edge7\">\n",
       "<title>remove_encode_blacklist-&gt;sort_masked_bams</title>\n",
       "<path d=\"M684.4598,-440.4567C667.8302,-433.7149 645.3934,-424.619 627.4089,-417.3279\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"627.914,-414.8891 620.5063,-414.5296 626.073,-419.4301 627.914,-414.8891\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"678\" y=\"-425.6\">input_file</text>\n",
       "</g>\n",
       "<!-- remove_encode_blacklist&#45;&gt;masked_file_basename -->\n",
       "<g class=\"edge\" id=\"edge19\">\n",
       "<title>remove_encode_blacklist-&gt;masked_file_basename</title>\n",
       "<path d=\"M708.8483,-440.4567C709.3362,-434.9678 709.9628,-427.9184 710.5286,-421.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"712.9734,-421.719 711.1529,-414.5296 708.0927,-421.2851 712.9734,-421.719\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"726\" y=\"-425.6\">input_file</text>\n",
       "</g>\n",
       "<!-- percent_mitochondrial_reads&#45;&gt;output_percent_mitochondrial_reads -->\n",
       "<g class=\"edge\" id=\"edge51\">\n",
       "<title>percent_mitochondrial_reads-&gt;output_percent_mitochondrial_reads</title>\n",
       "<path d=\"M851,-80.4567C851,-74.9678 851,-67.9184 851,-61.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"853.4501,-61.5296 851,-54.5296 848.5501,-61.5296 853.4501,-61.5296\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- mapped_reads_count&#45;&gt;output_read_count_mapped -->\n",
       "<g class=\"edge\" id=\"edge46\">\n",
       "<title>mapped_reads_count-&gt;output_read_count_mapped</title>\n",
       "<path d=\"M1206.7808,-80.4755C1216.9336,-76.945 1228.889,-73.2557 1240,-71 1281.5246,-62.57 1293.2003,-69.9386 1335,-63 1345.3275,-61.2857 1356.3522,-58.8464 1366.5911,-56.315\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1367.352,-58.6497 1373.5359,-54.5555 1366.1486,-53.8998 1367.352,-58.6497\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- percent_uniq_reads&#45;&gt;output_percentage_uniq_reads -->\n",
       "<g class=\"edge\" id=\"edge52\">\n",
       "<title>percent_uniq_reads-&gt;output_percentage_uniq_reads</title>\n",
       "<path d=\"M1070.8762,-80.4567C1078.5977,-74.2519 1088.8008,-66.0529 1097.4505,-59.1023\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"1099.219,-60.8242 1103.141,-54.5296 1096.1497,-57.0046 1099.219,-60.8242\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "<!-- mapped_filtered_reads_count&#45;&gt;output_read_count_mapped_filtered -->\n",
       "<g class=\"edge\" id=\"edge50\">\n",
       "<title>mapped_filtered_reads_count-&gt;output_read_count_mapped_filtered</title>\n",
       "<path d=\"M335.4861,-80.4567C332.1769,-74.7292 327.8859,-67.3025 324.0867,-60.727\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"326.1293,-59.365 320.506,-54.5296 321.8866,-61.8164 326.1293,-59.365\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg height=\"191pt\" viewBox=\"0.00 0.00 119.00 196.00\" width=\"116pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g class=\"graph\" id=\"graph0\" transform=\"scale(.9722 .9722) rotate(0) translate(4 192)\">\n",
       "<title>workflow</title>\n",
       "<g class=\"cluster\" id=\"clust1\">\n",
       "<title>cluster_inputs</title>\n",
       "<polygon fill=\"transparent\" points=\"8,-125 8,-180 93,-180 93,-125 8,-125\" stroke=\"#000000\" stroke-dasharray=\"5,2\"/>\n",
       "<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"50.5\" y=\"-168\">Workflow Inputs</text>\n",
       "</g>\n",
       "<g class=\"cluster\" id=\"clust2\">\n",
       "<title>cluster_outputs</title>\n",
       "<polygon fill=\"transparent\" points=\"11,-8 11,-63 103,-63 103,-8 11,-8\" stroke=\"#000000\" stroke-dasharray=\"5,2\"/>\n",
       "<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"57\" y=\"-15\">Workflow Outputs</text>\n",
       "</g>\n",
       "<!-- message -->\n",
       "<g class=\"node\" id=\"node1\">\n",
       "<title>message</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"22,-133.5 22,-152.5 78,-152.5 78,-133.5 22,-133.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"50\" y=\"-140.5\">message</text>\n",
       "</g>\n",
       "<!-- step0 -->\n",
       "<g class=\"node\" id=\"node3\">\n",
       "<title>step0</title>\n",
       "<polygon fill=\"#fafad2\" points=\"29.5,-80.5 29.5,-99.5 70.5,-99.5 70.5,-80.5 29.5,-80.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"50\" y=\"-87.5\">step0</text>\n",
       "</g>\n",
       "<!-- message&#45;&gt;step0 -->\n",
       "<g class=\"edge\" id=\"edge1\">\n",
       "<title>message-&gt;step0</title>\n",
       "<path d=\"M50,-133.279C50,-125.8827 50,-115.5239 50,-106.7951\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"52.4501,-106.6594 50,-99.6594 47.5501,-106.6594 52.4501,-106.6594\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"8.00\" text-anchor=\"middle\" x=\"66.5\" y=\"-110.6\">message</text>\n",
       "</g>\n",
       "<!-- response -->\n",
       "<g class=\"node\" id=\"node2\">\n",
       "<title>response</title>\n",
       "<polygon fill=\"#94ddf4\" points=\"22,-35.5 22,-54.5 78,-54.5 78,-35.5 22,-35.5\" stroke=\"#000000\"/>\n",
       "<text fill=\"#000000\" font-family=\"Helvetica,sans-Serif\" font-size=\"10.00\" text-anchor=\"middle\" x=\"50\" y=\"-42.5\">response</text>\n",
       "</g>\n",
       "<!-- step0&#45;&gt;response -->\n",
       "<g class=\"edge\" id=\"edge2\">\n",
       "<title>step0-&gt;response</title>\n",
       "<path d=\"M50,-80.4567C50,-74.9678 50,-67.9184 50,-61.5532\" fill=\"none\" stroke=\"#000000\"/>\n",
       "<polygon fill=\"#000000\" points=\"52.4501,-61.5296 50,-54.5296 47.5501,-61.5296 52.4501,-61.5296\" stroke=\"#000000\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import cwlBrowser as c\n",
    "import workflow\n",
    "import requests\n",
    "import yaml as y \n",
    "from IPython.display import SVG, display\n",
    "c.displayGraphSVG(\"https://github.com/alexbarrera/GGR-cwl/blob/master/v1.0/ATAC-seq_pipeline/03-map-se-blacklist-removal.cwl\")\n",
    "c.displayGraphSVG(\"https://github.com/seanleken/workflowtest/blob/master/workflows/newhello.cwl\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validating my own created workflows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'cwlVersion': 'v1.0', 'class': 'Workflow', 'label': 'Hello World', 'doc': 'Outputs a message using echo', 'inputs': {'message': {'type': 'string', 'doc': 'The message to print', 'default': 'Hello World', 'inputBinding': {'position': 0}}}, 'outputs': {'response': {'outputSource': 'step0/response', 'type': 'File'}}, 'steps': {'step0': {'run': {'class': 'CommandLineTool', 'inputs': {'message': {'type': 'string'}}, 'baseCommand': 'echo', 'stdout': 'response.txt', 'outputs': {'response': {'type': 'stdout'}}}, 'in': {'message': 'message'}, 'out': ['response']}}}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import base64\n",
    "import yaml\n",
    "\n",
    "link = \"https://api.github.com/repos/seanleken/workflowtest/contents/workflows/newhello.cwl\"\n",
    "req = requests.get(link)\n",
    "if req.status_code != 200:\n",
    "    print(req.status_code)\n",
    "    if req.status_code == 404:\n",
    "        print (\"cwl file not found. Maybe due to incorrect url\")\n",
    "        print(\"\\n\")\n",
    "else:\n",
    "    req = req.json()  \t    \n",
    "    content = base64.b64decode(req['content'])\n",
    "    content = content.decode(\"utf-8\")\n",
    "    try:\n",
    "        workflow = yaml.safe_load(content)\n",
    "        print (workflow)\n",
    "    except (yaml.YAMLError) as yamlError:\n",
    "        print (\"Error in loading the cwl file\")\n",
    "        print (yamlError)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Similarity check test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SIMILARITY CHECK BETWEEN lobSTR.cwl AND lobSTR.cwl:\n",
      "---------------------------------------------------------------------------------------\n",
      "OVERALL MATCH: 100.0\n",
      "---------------------------------------------------------------------------------------\n",
      "\n",
      "\n",
      "SIMILARITY CHECK BETWEEN lobSTR.cwl AND lobSTR1.cwl:\n",
      "THE FOLLOWING ARE THE STEPS THAT DIFFER\n",
      "----------------------------------------------------------\n",
      "lobSTR1.cwl\n",
      "-----------------------------------------------------------\n",
      "samsort5\n",
      "\n",
      "lobSTR.cwl\n",
      "-----------------------------------------------------------\n",
      "samsort\n",
      "\n",
      "\n",
      "---------------------------------------------------------------------------------------\n",
      "OVERALL MATCH: 82.5\n",
      "---------------------------------------------------------------------------------------\n",
      "\n",
      "\n",
      "SIMILARITY CHECK BETWEEN lobSTR.cwl AND lobSTR3.cwl:\n",
      "THE FOLLOWING ARE THE INPUTS THAT DIFFER\n",
      "----------------------------------------------------------\n",
      "lobSTR.cwl\n",
      "-----------------------------------------------------------\n",
      "\n",
      "lobSTR3.cwl\n",
      "-----------------------------------------------------------\n",
      "p3\n",
      "\n",
      "\n",
      "---------------------------------------------------------------------------------------\n",
      "OVERALL MATCH: 98.33333333333333\n",
      "---------------------------------------------------------------------------------------\n",
      "\n",
      "\n",
      "SIMILARITY CHECK BETWEEN lobSTR.cwl AND mixed_library_metrics.cwl:\n",
      "THE FOLLOWING ARE THE STEPS THAT DIFFER\n",
      "----------------------------------------------------------\n",
      "lobSTR.cwl\n",
      "-----------------------------------------------------------\n",
      "samindex\n",
      "lobSTR\n",
      "samsort\n",
      "allelotype\n",
      "\n",
      "mixed_library_metrics.cwl\n",
      "-----------------------------------------------------------\n",
      "samtools_flagstat_to_sqlite\n",
      "samtools_flagstat\n",
      "picard_collectoxogmetrics\n",
      "samtools_idxstats_to_sqlite\n",
      "picard_collectmultiplemetrics\n",
      "picard_collectwgsmetrics_to_sqlite\n",
      "samtools_stats\n",
      "merge_sqlite\n",
      "picard_collectoxogmetrics_to_sqlite\n",
      "samtools_idxstats\n",
      "picard_collectmultiplemetrics_to_sqlite\n",
      "samtools_stats_to_sqlite\n",
      "picard_collectwgsmetrics\n",
      "\n",
      "\n",
      "THE FOLLOWING ARE THE INPUTS THAT DIFFER\n",
      "----------------------------------------------------------\n",
      "mixed_library_metrics.cwl\n",
      "-----------------------------------------------------------\n",
      "db_snp_vcf\n",
      "bam\n",
      "thread_count\n",
      "fasta\n",
      "input_state\n",
      "uuid\n",
      "\n",
      "lobSTR.cwl\n",
      "-----------------------------------------------------------\n",
      "p1\n",
      "output_prefix\n",
      "p2\n",
      "strinfo\n",
      "noise_model\n",
      "rg-sample\n",
      "rg-lib\n",
      "reference\n",
      "\n",
      "\n",
      "THE FOLLOWING ARE THE OUTPUTS THAT DIFFER\n",
      "----------------------------------------------------------\n",
      "mixed_library_metrics.cwl\n",
      "-----------------------------------------------------------\n",
      "merge_sqlite_destination_sqlite\n",
      "\n",
      "lobSTR.cwl\n",
      "-----------------------------------------------------------\n",
      "bam\n",
      "vcf\n",
      "bam_stats\n",
      "vcf_stats\n",
      "\n",
      "\n",
      "---------------------------------------------------------------------------------------\n",
      "OVERALL MATCH: 0.0\n",
      "---------------------------------------------------------------------------------------\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import cwlBrowser as c\n",
    "import workflow\n",
    "import requests\n",
    "import yaml as y\n",
    "\n",
    "lobSTR1 = c.load(\"lobSTR.cwl\")\n",
    "lobSTR2 = c.load(\"lobSTR.cwl\")\n",
    "lobSTR3 = c.load(\"lobSTR1.cwl\")\n",
    "lobSTR4 = c.load(\"lobSTR3.cwl\")\n",
    "metrics = c.load(\"mixed_library_metrics.cwl\")\n",
    "c.similarityCheck(lobSTR1, lobSTR2)\n",
    "c.similarityCheck(lobSTR1, lobSTR3)\n",
    "c.similarityCheck(lobSTR1, lobSTR4)\n",
    "c.similarityCheck(lobSTR1, metrics)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
